### PR TITLE
New command devices list

### PR DESCRIPTION
**Pull request by lluiscampos**
* New command devices list
Fixed minor details in existing code for coherence. Adding acceptance
tests also.
Changelog: New command `mender-cli devices list` to list all devices
from /devauth/devices endpoint. The amount of detail can be controlled
using cli parameter `-d/--detail`, same as for other commands.

* [http clients code] Refactor common code into common.go

### CODE REVIEW

**Comment by oleorhagen** [comment]
Actually, @lluiscampos could you review quickly https://github.com/mendersoftware/mender-cli/pull/112 and rebase this on top of it, so that we don't let it slip a second time?

File: client/devices/client.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+	return &Client{
+		url:            url,
+		devicesListURL: JoinURL(url, devicesListURL),
+		client: &http.Client{
+			Transport: tr,
+		},
+	}
+}
+
+func (c *Client) ListDevices(tokenPath string, detailLevel int) error {
```
Comment (oleorhagen) [inline client/devices/client.go:82]: Musing:
with the amount of code shared with `/deployments/client` it should be possible to abstract away most of the HTTP request creation. OTOH, I don't really see a huge benefit to it, what do you think?

File: client/devices/client.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+			fmt.Printf("  SN: %s\n", a.IdentityData.Sn)
+		}
+	}
+	if detailLevel >= 1 {
+		fmt.Printf("CreatedTs: %s\n", a.CreatedTs)
+		fmt.Printf("UpdatedTs: %s\n", a.UpdatedTs)
+		fmt.Printf("Decommissioning: %t\n", a.Decommissioning)
+	}
+	if detailLevel >= 2 {
+		for i, v := range a.AuthSets {
```
Comment (oleorhagen) [inline client/devices/client.go:151]: Feel free to ignore:
Could be a nice use-case for `text-templates` I think

File: client/devices/client.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+			}
+			fmt.Printf("  Status: %s\n", v.Status)
+			fmt.Printf("  Ts: %s\n", v.Ts)
+		}
+	}
+
+	fmt.Println("--------------------------------------------------------------------------------")
+}
+
+func JoinURL(base, url string) string {
```
Comment (oleorhagen) [inline client/devices/client.go:173]: Suggestion:
How about `path.Join` ?

File: cmd/devices_list.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+package cmd
+
+import (
+	"github.com/spf13/cobra"
+
+	"github.com/mendersoftware/mender-cli/client/devices"
+)
+
+// const (
+// 	argDetailLevel = "detail"
```
Comment (oleorhagen) [inline cmd/devices_list.go:23]: Leftover

File: cmd/devices_list.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+// )
+
+var devicesListCmd = &cobra.Command{
+	Use:   "list",
+	Short: "List Mender Devices from the Mender server.",
+	Run: func(c *cobra.Command, args []string) {
+		cmd, err := NewDevicesListCmd(c, args)
+		CheckErr(err)
+
```
Comment (oleorhagen) [inline cmd/devices_list.go:32]: Question:
Why empty air here?

File: cmd/devices_list.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+type DevicesListCmd struct {
+	server      string
+	skipVerify  bool
+	tokenPath   string
+	detailLevel int
+}
+
+func NewDevicesListCmd(cmd *cobra.Command, args []string) (*DevicesListCmd, error) {
+	server, err := cmd.Flags().GetString(argRootServer)
```
Comment (oleorhagen) [inline cmd/devices_list.go:49]:
```suggestion
	server  := viper.GetString(argRootServer)
```

File: cmd/devices_list.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+	if err != nil {
```
Comment (oleorhagen) [inline cmd/devices_list.go:50]:
```suggestion
	if server == "" {
```

File: cmd/devices_list.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+		tokenPath:   token,
+		skipVerify:  skipVerify,
+		detailLevel: detailLevel,
+	}, nil
+}
+
+func (c *DevicesListCmd) Run() error {
+
+	client := devices.NewClient(c.server, c.skipVerify)
+	err := client.ListDevices(c.tokenPath, c.detailLevel)
```
Comment (oleorhagen) [inline cmd/devices_list.go:87]:
```suggestion
	return client.ListDevices(c.tokenPath, c.detailLevel)
```

Comment (oleorhagen) [reply cmd/devices_list.go:87]: Since we're not really using the returned error here

File: tests/tests/test_devices.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+        "https://mender-api-gateway",
+        "--skip-verify",
+        "--username",
+        "user@tenant.com",
+        "--password",
+        "youcantguess",
+    )
+
+    assert r.returncode == 0, r.stderr
+    yield
```
Comment (oleorhagen) [inline tests/tests/test_devices.py:42]: Teardowns are out of fashion? :p

File: tests/tests/test_devices.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+        r = mapi.set_device_auth_status(device_id, authset_id, "accepted")
+        assert r.status_code == 204
+
+        # Device should now be listed as accepted
+        c = cli.Cli()
+        r = c.run(
+            "--server", "https://mender-api-gateway", "--skip-verify", "devices", "list"
+        )
+        assert r.returncode == 0, r.stderr
+        expect_output(r.stdout, "Status: accepted")
```
Comment (oleorhagen) [inline tests/tests/test_devices.py:95]: Question:
With one input, this is really just a regular `assert`, right?

**Review by oleorhagen — APPROVED** [review]
Real nice!

**Comment by oleorhagen** [comment]
Remember rebase

File: client/devices/client.go [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+type devicesList struct {
+	devices []deviceData
+}
+
+type deviceData struct {
+	ID           string `json:"id"`
+	IdentityData struct {
+		Mac string `json:"mac"`
+		Sku string `json:"sku"`
+		Sn  string `json:"sn"`
```
Comment (merlin-northern) [inline client/devices/client.go:39]: Mac Sku Sn -- can we be more verbose with the naming?

File: client/devices/client.go [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+			}
+			if v.IdentityData.Sn != "" {
+				fmt.Printf("    SN: %s\n", v.IdentityData.Sn)
+			}
+			fmt.Printf("  Status: %s\n", v.Status)
+			fmt.Printf("  Ts: %s\n", v.Ts)
+		}
+	}
+
+	fmt.Println("--------------------------------------------------------------------------------")
```
Comment (merlin-northern) [inline client/devices/client.go:170]: realy nitpicky and mean: what if my terminal is less than 80 chars in width?

File: cmd/devices_list.go [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+	"github.com/mendersoftware/mender-cli/client/devices"
+)
+
+// const (
+// 	argDetailLevel = "detail"
+// )
+
+var devicesListCmd = &cobra.Command{
+	Use:   "list",
+	Short: "List Mender Devices from the Mender server.",
```
Comment (merlin-northern) [inline cmd/devices_list.go:28]: maybe rephrase a bit? suggested: "Get a list of devices from the Mender server"

File: cmd/devices_list.go [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+
+func init() {
+	devicesListCmd.Flags().IntP(argDetailLevel, "d", 0, "devices list detail level [0..3]")
+}
+
+type DevicesListCmd struct {
+	server      string
+	skipVerify  bool
+	tokenPath   string
+	detailLevel int
```
Comment (merlin-northern) [inline cmd/devices_list.go:45]: minor/maybe stupid: can we name 0 1 2 3 detail levels? as in: Normal Verbose ...

**Review by merlin-northern — APPROVED** [review]
lgtm left some comments.

File: client/devices/client.go (hunk 3 above)
Comment (lluiscampos) [reply client/devices/client.go:173]: Thanks. Cleaned up also from `useradm` and `deployments`

File: cmd/devices_list.go (hunk 5 above)
Comment (lluiscampos) [reply cmd/devices_list.go:32]: Cleaned also in all other files

File: client/devices/client.go (hunk 3 above)
Comment (lluiscampos) [reply client/devices/client.go:173]: Nah, `path.Join` does not work nicely with URLs, really. I'll leave as is but moving it to a common place.

File: client/devices/client.go (hunk 1 above)
Comment (lluiscampos) [reply client/devices/client.go:82]: I'll try to refactor a bit :)

File: cmd/devices_list.go (hunk 8 above)
Comment (lluiscampos) [reply cmd/devices_list.go:87]: Fixed also in `artifacts_list`

File: tests/tests/test_devices.py (hunk 10 above)
Comment (lluiscampos) [reply tests/tests/test_devices.py:95]: It is, but just for reusing the existing code

File: client/devices/client.go (hunk 2 above)
Comment (lluiscampos) [reply client/devices/client.go:151]: Maybe next time... I have to learn them first :)

File: client/devices/client.go (hunk 12 above)
Comment (lluiscampos) [reply client/devices/client.go:170]: I'll leave as-is: same behavior as for artifacts list. I'll think about a common fix later

File: cmd/devices_list.go (hunk 14 above)
Comment (lluiscampos) [reply cmd/devices_list.go:45]: Same here: I'll leave it for a further rework (together with artifacts list)

File: cmd/devices_list.go (hunk 13 above)
Comment (lluiscampos) [reply cmd/devices_list.go:28]: Fixed also similar sentence in artifacts_list

File: client/devices/client.go (hunk 11 above)
Comment (lluiscampos) [reply client/devices/client.go:39]: What are they? I followed the naming from API docs

Comment (lluiscampos) [reply client/devices/client.go:39]: Found it, I'll use the full name from:

image

**Comment by lluiscampos** [comment]
Amended following most of your feedback and refactored a bit the common code.

**Comment by lluiscampos** [comment]
> Amended following most of your feedback and refactored a bit the common code.

(and rebased, of course)

**Comment by oleorhagen** [comment]
Looks real good
